File: python/tests/test_scan.py
"""Tests for the duplicate scanner."""

from __future__ import annotations

import json
from pathlib import Path

from typer.testing import CliRunner

from xlsreport import __version__
from xlsreport.__main__ import app
from xlsreport.scan import scan

runner = CliRunner()


def build_tree(root: Path) -> None:
    """Three copies of A, two of B, one unique, one unique-by-size."""
    (root / "nested" / "deep").mkdir(parents=True)
    (root / "a1.txt").write_bytes(b"AAAA")
    (root / "nested" / "a2.txt").write_bytes(b"AAAA")
    (root / "nested" / "deep" / "a3.txt").write_bytes(b"AAAA")
    (root / "b1.bin").write_bytes(b"BBBB")
    (root / "nested" / "b2.bin").write_bytes(b"BBBB")
    (root / "c1.log").write_bytes(b"CCCC")
    (root / "nested" / "deep" / "d1.dat").write_bytes(b"DDDDDDDDDD")


def test_version() -> None:
    assert __version__ == "0.1.0"


def test_scan_counts(tmp_path: Path) -> None:
    build_tree(tmp_path)
    result = scan(str(tmp_path))
    assert result.total_files == 7
    assert result.total_bytes == 4 * 6 + 10
    # A(3) + B(2) + C(1) + D(1)
    assert result.distinct_contents == 4
    assert result.duplicate_groups == 2
    assert result.redundant_copies == 3
    assert result.redundant_bytes == 4 * 3


def test_scan_skips_unique_sizes(tmp_path: Path) -> None:
    """A size seen once is never hashed."""
    build_tree(tmp_path)
    result = scan(str(tmp_path))
    # d1.dat is the only 10-byte file, so it is skipped by the size filter.
    assert result.hashed_files == 6


def test_empty_tree(tmp_path: Path) -> None:
    result = scan(str(tmp_path))
    assert result.total_files == 0
    assert result.duplicate_groups == 0
    assert result.distinct_contents == 0


def test_cli_json(tmp_path: Path) -> None:
    build_tree(tmp_path)
    outcome = runner.invoke(app, ["scan", str(tmp_path), "--json"])
    assert outcome.exit_code == 0
    payload = json.loads(outcome.stdout)
    assert payload["tool"] == "xlsreport"
    assert payload["total_files"] == 7
    assert payload["duplicate_groups"] == 2
    assert payload["redundant_copies"] == 3

File: cpp/tests/test_scan.cpp
#include <gtest/gtest.h>

#include <filesystem>
#include <fstream>
#include <string>

#include "triage/scan.hpp"

namespace fs = std::filesystem;

namespace {

void write_file(const fs::path& path, const std::string& content) {
  fs::create_directories(path.parent_path());
  std::ofstream out(path, std::ios::binary);
  out << content;
}

// Mirrors build_tree() in python/tests/test_scan.py so both suites assert
// the same numbers against the same shape of input.
fs::path build_tree() {
  const fs::path root = fs::temp_directory_path() / "triage_scan_test";
  fs::remove_all(root);
  fs::create_directories(root / "nested" / "deep");
  write_file(root / "a1.txt", "AAAA");
  write_file(root / "nested" / "a2.txt", "AAAA");
  write_file(root / "nested" / "deep" / "a3.txt", "AAAA");
  write_file(root / "b1.bin", "BBBB");
  write_file(root / "nested" / "b2.bin", "BBBB");
  write_file(root / "c1.log", "CCCC");
  write_file(root / "nested" / "deep" / "d1.dat", "DDDDDDDDDD");
  return root;
}

}  // namespace

TEST(ScanTest, CountsMatchKnownTree) {
  const fs::path root = build_tree();
  const triage::ScanResult result = triage::scan(root.string());

  EXPECT_EQ(result.total_files, 7u);
  EXPECT_EQ(result.total_bytes, (4u * 6u) + 10u);
  EXPECT_EQ(result.distinct_contents, 4u);
  EXPECT_EQ(result.duplicate_groups, 2u);
  EXPECT_EQ(result.redundant_copies, 3u);
  EXPECT_EQ(result.redundant_bytes, 12u);

  fs::remove_all(root);
}

TEST(ScanTest, SkipsSizesSeenOnce) {
  const fs::path root = build_tree();
  const triage::ScanResult result = triage::scan(root.string());
  // d1.dat is the only 10-byte file, so the size filter drops it before hashing.
  EXPECT_EQ(result.hashed_files, 6u);
  fs::remove_all(root);
}

TEST(ScanTest, EmptyTree) {
  const fs::path root = fs::temp_directory_path() / "triage_scan_empty";
  fs::remove_all(root);
  fs::create_directories(root);

  const triage::ScanResult result = triage::scan(root.string());
  EXPECT_EQ(result.total_files, 0u);
  EXPECT_EQ(result.duplicate_groups, 0u);
  EXPECT_EQ(result.distinct_contents, 0u);

  fs::remove_all(root);
}

TEST(ScanTest, JsonContainsToolName) {
  const triage::ScanResult result;
  const std::string json = triage::to_json(result);
  EXPECT_NE(json.find("\"tool\": \"triage\""), std::string::npos);
}
